### DISCUSSION TITLE
dbt snowflake application parameter

### DISCUSSION THREAD

**Discussion by o-shevchenko** [dbt-labs/dbt-snowflake#254] [Q&A] [answered]
Hardcoded `application="dbt"`parameter for Snowflake connection makes it impossible to use dbt sometimes. Why it's not possible to override? Can we open PR to make it configurable?
https://github.com/dbt-labs/dbt-snowflake/blob/main/dbt/adapters/snowflake/connections.py#L287

**Comment by jtcohen6** [comment, marked as answer]
> makes it impossible to use dbt sometimes

@o-shevchenko Could you say more here?

We've discussed this in the past (https://github.com/dbt-labs/dbt-snowflake/issues/148), and we're unlikely to change it, but I'm always open to hearing that there's a specific use case we haven't considered

**Reply by o-shevchenko** [reply]
We have requirements for our work with Snowflake, so we can differentiate our workload on Snowflake from our organization. We probably can just fork Snowflake connector and hardcode our organization but instead of that, we want just to be able to configure it. Is there any reason to hardcode "dbt"? I think it makes more sense just make "dbt" as the default value and allow to override it if needed.

**Reply by o-shevchenko** [reply]
For example, we use "application" parameter for Spark Snowflake DataSource for the same reason.

**Reply by o-shevchenko** [reply]
Not sure if "query_tag" can cover partner usage and that's what we really need.

**Reply by o-shevchenko** [reply]
https://docs.snowflake.com/en/user-guide/odbc-parameters.html
<img width="1327" alt="image" src="https://user-images.githubusercontent.com/31073930/189366995-53a8681b-f0cf-4872-b7fb-8267146f623b.png">

**Reply by o-shevchenko** [reply]
See also https://docs.snowflake.com/en/user-guide/ecosystem-partner-connect.html

**Reply by o-shevchenko** [reply]
@jtcohen6 Any thoughts?

**Reply by jtcohen6** [reply]
Hey @o-shevchenko - dbt is also a Snowflake partner, and as I understand it, it's an important part of our agreement with them that all Snowflake queries orchestrated by dbt be attributable to `dbt`. That's true whether it's a specific organization's use of dbt for their own analytics, or if `dbt-snowflake` is being wrapped in and executed by another tool.

I hate to say "fork," but this isn't something that we're going to make configurable in the official `dbt-snowflake` adapter.

For what it's worth, this feels like a conversation that might better be had with the dbt Labs Partnership team. We just launched our program around technology partners; you can read more about it [here](https://www.getdbt.com/blog/dbt-labs-technology-partner-program/), and check out the portal for partners [here](https://partners.getdbt.com/English/).

**Reply by o-shevchenko** [reply]
That's an interesting topic. I understand that dbt Labs as Snowflake partner wants to gain credits by using their own account, but it looks strange for open source software.  It's ok for dbt Cloud but not for software with Apache 2.0 license. Other Snowflake partners won't be able to use such software as is, but they can fork it and make changes to fix such a problem.  Or will use a different tool for their workload instead of dbt.  So, not sure I understand why it's not allowed to make "application" configurable since it doesn't make sense. Do you think it will reduce the usage of "dbt" application dramatically even if you still use "dbt" by default but make it configurable?

**Reply by jtcohen6** [reply]
@o-shevchenko I agree it's an interesting topic! This conversation has come up a few times before (though only a few), and it's possible that the answer will be different at some point in the future. For now, the answer is no.

In my view, it's important to the entire dbt community — not just dbt Labs as a company or commercial parter of Snowflake — that dbt gets credit where due. That is, whenever people use the official dbt adapter for Snowflake, in whatever context, it always shows up as `dbt`. This enables us to punch at our collective weight. It means Snowflake takes the dbt community seriously, and takes seriously dbt as the crucial piece of open source software that we all rely on to get the most out of what Snowflake has to offer.

> It's ok for dbt Cloud but not for software with Apache 2.0 license.

I see where you’re coming from, but I disagree. For one thing, we don’t differentiate between dbt Cloud and dbt Core for attributing this usage: it’s always `dbt`.

At the risk of kicking off an abstract philosophical discussion about the spirit of open source software… Anyone in the world can use `dbt-core`, and the `dbt-snowflake` adapter plugin, for free and for any purpose. There's significant work that goes into building and maintaining these as pieces of open source software, including the work of reviewing and accepting open source contributions — all for the benefit of the larger community of OSS users. In that spirit, I think `dbt` should get credit where due. So for anyone who uses the official `dbt-snowflake` adapter plugin to orchestrate their Snowflake queries — whether wrapped in another partner application or not — Snowflake should know that their queries are being most directly orchestrated by dbt, and their usage should be counted among the larger community’s.<sup>1</sup>

Of course, because `dbt-snowflake` has an Apache 2.0 license, you’re within your rights to fork it and make changes. If it’s just to change that one line — to redirect credit, to opt out of the larger collective — I can’t stop you. If it’s also to make meaningful improvements, other customizations — that's the spirit! We’d be thrilled if you decided to contribute some of those changes back upstream :)

<sup>1</sup>As a total aside: Writing this sentence reminded me of a song I liked when I was younger: “To Be Counted Among Men,” by the Irish band Villagers
